sql/src/main/java/io/crate/operation/NodeJobsTracker.java

```diff
@@ -0,0 +1,110 @@
+/*
+ * Licensed to Crate under one or more contributor license agreements.
+ * See the NOTICE file distributed with this work for additional
+ * information regarding copyright ownership.  Crate licenses this file
+ * to you under the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.  You may
+ * obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+ * implied.  See the License for the specific language governing
+ * permissions and limitations under the License.
+ *
+ * However, if you have executed another commercial license agreement
+ * with Crate these terms will supersede the license and you may use the
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.operation;
+
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.common.inject.Singleton;
+import org.elasticsearch.common.logging.Loggers;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Map;
+import java.util.UUID;
+import java.util.concurrent.ConcurrentHashMap;
+
+/**
+ * Monitors how many operations issued from the current node are in progress across the cluster.
+ * Note: one job can span multiple nodes.
+ */
+@Singleton
+public class NodeJobsTracker {
+
+    private static final Logger logger = Loggers.getLogger(NodeJobsTracker.class);
+
+    private final Map<String, Long> operationsCountPerNode = new ConcurrentHashMap<>();
+    private final Map<UUID, List<String>> jobToNodesMap = new ConcurrentHashMap<>();
```

Comment:
Do we need jobId awareness here? I think it would be more efficient to move this closer to the callsite where it might be possible to use a data structure that doesn't have to be thread-safe